### PR TITLE
initialization of Junit 5

### PR DESCRIPTION
**Pull request by DaliborSkrobak**
Hi guys,

I was looking on test coverage and I thought that I will improve it little bit, by writing new JUnit tests.
Then I realized that it would be perfect way to learn/use new tricks from JUnit5.

So I thought that it would make sense to first to update OpenGrok to using JUnit 5 and then start changing/adding tests.

What do you think?

Tested by:
```bash
./mvnw test
...

========================== 54 passed in 18.41 seconds ==========================
[INFO] 
[INFO] ---------------------< org.opengrok:opengrok-dist >---------------------
[INFO] Building Distribution 1.2.8                                        [7/7]
[INFO] --------------------------------[ pom ]---------------------------------
[INFO] 
[INFO] --- maven-enforcer-plugin:3.0.0-M2:enforce (enforce-maven) @ opengrok-dist ---
[INFO] 
[INFO] --- jacoco-maven-plugin:0.8.2:prepare-agent (pre-unit-test) @ opengrok-dist ---
[INFO] surefireArgLine set to -javaagent:/home/dalibor/.m2/repository/org/jacoco/org.jacoco.agent/0.8.2/org.jacoco.agent-0.8.2-runtime.jar=destfile=/media/truecrypt1/repo/github/opengrok/distribution/target/jacoco.exec
[INFO] 
[INFO] --- maven-surefire-plugin:3.0.0-M3:test (run-unit-tests) @ opengrok-dist ---
[INFO] No tests to run.
[INFO] 
[INFO] --- jacoco-maven-plugin:0.8.2:report (post-unit-test) @ opengrok-dist ---
[INFO] Skipping JaCoCo execution due to missing execution data file.
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for OpenGrok 1.2.8:
[INFO] 
[INFO] OpenGrok ........................................... SUCCESS [  1.796 s]
[INFO] OpenGrok Indexer ................................... SUCCESS [ 19.060 s]
[INFO] OpenGrok authorization plugins ..................... SUCCESS [  0.540 s]
[INFO] OpenGrok Suggester ................................. SUCCESS [ 53.471 s]
[INFO] OpenGrok Web ....................................... SUCCESS [03:19 min]
[INFO] OpenGrok tools ..................................... SUCCESS [ 24.490 s]
[INFO] Distribution ....................................... SUCCESS [  0.040 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  04:59 min
[INFO] Finished at: 2019-06-01T01:30:26-07:00
[INFO] ------------------------------------------------------------------------

```

Thanks,
D.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -55,8 +55,7 @@ pmd.bat
 .project
 .settings/
 .vscode/
-*/.project
-*/.settings
-*/.classpath
```
Comment (vladak) [inline .gitignore]: is this change intentional ?

**Comment by vladak** [comment]
In general no problem with this. Could you shed more light on the 2 engines ?

File: .gitignore (hunk 1 above)
Comment (DaliborSkrobak) [reply .gitignore]: yes it is. I'm trying to set up my eclipse and I'm seeing some files which should not be tracked (`*/.` covers only subdirectories, not files in root git dir)

Below is example without my changes in `.gitignore` file:
```
# git status --short

# touch .classpath
# touch opengrok-indexer/.classpath
# git status --short
?? .classpath
```
`.classpath` file is not shown in subdirectory, but it is visible in root directory.

BTW [.gitignore man page](https://git-scm.com/docs/gitignore):
>If the pattern does not contain a slash /, Git treats it as a shell glob pattern and checks for a match 
>against the pathname relative to the location of the .gitignore file (relative to the toplevel of the work tree 
>if not from a .gitignore file).

**Comment by DaliborSkrobak** [comment]
So old way (JUnit 3/4) contained one JAR file which had everything in it.  JUnit 5 was spitted into three parts/engines:

from [JUnit 5 user guide](https://junit.org/junit5/docs/current/user-guide/):
> JUnit 5 = JUnit Platform + JUnit Jupiter + JUnit Vintage
> 
>The JUnit Platform serves as a foundation for launching testing frameworks on the JVM. It also defines the TestEngine API for developing a testing framework that runs on the platform. Furthermore, the platform provides a Console Launcher to launch the platform from the command line and a JUnit 4 based Runner for running any TestEngine on the platform in a JUnit 4 based environment. First-class support for the JUnit Platform also exists in popular IDEs (see IntelliJ IDEA, Eclipse, NetBeans, and Visual Studio Code) and build tools (see Gradle, Maven, and Ant).
>
>JUnit Jupiter is the combination of the new programming model and extension model for writing tests and extensions in JUnit 5. The Jupiter sub-project provides a TestEngine for running Jupiter based tests on the platform.
>
>JUnit Vintage provides a TestEngine for running JUnit 3 and JUnit 4 based tests on the platform.

So for OpenGrok `JUnit Jupiter` is needed used for writing JUnit5 tests. And `JUnit Vintage` is used to allow running already implemented tests.

**Comment by DaliborSkrobak** [comment]
There is something strange around `@ConditionalRun` classes. I'm closing PR to prevent integration of this change. I will reopen when I will now that everything is really fine

**Comment by vladak** [comment]
Okay, thanks !

Dne po 3. 6. 2019 10:00 uživatel DaliborSkrobak <notifications@github.com>
napsal:

**Comment by DaliborSkrobak** [comment]
tested by:
```
# git log --oneline -5
b0d7d61c (HEAD -> dalibor-junit5-init) shows full stacktrace for failed JUnit tests
57200fd4 updated maven-surefire plugin to detect JUnits fully
9eb55e65 Eclipse IDE file properly ignored
d2b87cd5 initialization of Junit 5
289556ea (origin/master, origin/HEAD, master) use wildcard instead of list all files and folders

# ./mvnw clean test surefire-report:report
...
[INFO] <<< maven-surefire-report-plugin:3.0.0-M3:report (default-cli) < [surefire]test @ opengrok-dist <<<
[INFO]
[INFO]
[INFO] --- maven-surefire-report-plugin:3.0.0-M3:report (default-cli) @ opengrok-dist ---
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for OpenGrok 1.2.8:
[INFO]
[INFO] OpenGrok ........................................... SUCCESS [  3.588 s]
[INFO] OpenGrok Indexer ................................... SUCCESS [11:17 min]
[INFO] OpenGrok authorization plugins ..................... SUCCESS [ 28.735 s]
[INFO] OpenGrok Suggester ................................. SUCCESS [ 50.106 s]
[INFO] OpenGrok Web ....................................... SUCCESS [03:36 min]
[INFO] OpenGrok tools ..................................... SUCCESS [01:11 min]
[INFO] Distribution ....................................... SUCCESS [  0.331 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  17:29 min
[INFO] Finished at: 2019-06-11T13:50:41-07:00
[INFO] ------------------------------------------------------------------------

# find . -type d -name surefire-reports | xargs ls  | grep txt$ | wc -l
219

# git co master

# ./mvnw clean

# find . -type d -name surefire-reports | xargs ls  | grep txt$ | wc -l
9

# ./mvnw clean test surefire-report:report
...
[INFO] <<< maven-surefire-report-plugin:3.0.0-M3:report (default-cli) < [surefire]test @ opengrok-dist <<<
[INFO]
[INFO]
[INFO] --- maven-surefire-report-plugin:3.0.0-M3:report (default-cli) @ opengrok-dist ---
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for OpenGrok 1.2.8:
[INFO]
[INFO] OpenGrok ........................................... SUCCESS [  2.543 s]
[INFO] OpenGrok Indexer ................................... SUCCESS [09:04 min]
[INFO] OpenGrok authorization plugins ..................... SUCCESS [ 21.968 s]
[INFO] OpenGrok Suggester ................................. SUCCESS [ 54.371 s]
[INFO] OpenGrok Web ....................................... SUCCESS [02:54 min]
[INFO] OpenGrok tools ..................................... SUCCESS [ 55.301 s]
[INFO] Distribution ....................................... SUCCESS [  0.144 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  14:13 min
[INFO] Finished at: 2019-06-11T14:07:20-07:00
[INFO] ------------------------------------------------------------------------

# find . -type d -name surefire-reports | xargs ls  | grep txt$ | wc -l
219
```

**Comment by vladak** [comment]
What was the problem with `@ConditionalRun` ?

**Comment by vladak** [comment]
I wonder if the duplicated dependencies for each module can be inherited somehow.

**Comment by tarzanek** [comment]
so new junit 5 is SLOWER than old one? wft?

**Comment by DaliborSkrobak** [comment]
> What was the problem with @ConditionalRun ?

I was wrong in the end. I noticed that some of JUnits were ignored and initially it looked like problem with @ConditionalRun, because I those tests were missing in my Eclipse. When I re-run it directly through maven cmd, I noticed that all of tests were skipped (not just ConditionalRuns).
It was combination of two things:
- junit vintage must be declared as dependency
- surefire plugin 2.22.0 does not detect properly JUnit5 tests (hence upgrade of that plugin is needed)

> I wonder if the duplicated dependencies for each module can be inherited somehow.

good point, I was thinking about make that dependency for all sub-projects, but decided to not to, because then all sub-project will have it (e.g. _distribution_ and _opengrok_-tools).

BTW difference is in way how parent _pom.xml_ is defined.
If dependency is defined inside of [_\<dependencyManagement\>_](https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#Dependency_Management) xml tag, then it allows to "preset" some information about dependency in parent project (e.g. version), but sub-projects must still declare them. If those dependencies are declared just in _\<depenendencies\>_ xml tag, then automatically all sub-projects will inherit it.

> so new junit 5 is SLOWER than old one? wft?

oh yes, but I would not be inclined that JUnit5 is slower than JUnit4. Maybe just _vintage_ module is little bit slower, because it somehow wraps its calls around JUnit4. 
BTW I was playing little bit with parallel execution and it looked promising - I have plan to do some more testing around it.

**Comment by vladak** [comment]
Thanks for the detailed explanation. And welcome back to the party ! :-)
